crates/prune/prune/src/segments/history.rs

```diff
@@ -10,103 +10,158 @@ use reth_db_api::{
+    let mut unchanged = 0;
+    let mut cursor = provider.tx_ref().cursor_write::<RawTable<T>>()?;
+
+    for sharded_key in highest_sharded_keys {
+        // Seek to the shard that has the key >= the given sharded key
+        // TODO: optimize
+        let result = cursor
+            .seek(RawKey::new(sharded_key.clone()))?
+            .map(|(key, value)| Result::<_, DatabaseError>::Ok((key.key()?, value)))
+            .transpose()?;
 
-    // Prune history table:
-    // 1. If the shard has `highest_block_number` less than or equal to the target block number
-    // for pruning, delete the shard completely.
-    // 2. If the shard has `highest_block_number` greater than the target block number for
-    // pruning, filter block numbers inside the shard which are less than the target
-    // block number for pruning.
-    while let Some(result) = cursor.next()? {
-        let (key, blocks): (T::Key, BlockNumberList) = result;
+        // If such shard doesn't exist, skip to the next sharded key
+        if result.as_ref().map_or(true, |(key, _)| !key_matches(key, &sharded_key)) {
+            continue
+        }
+
+        // At this point, we're sure that the shard with the given sharded key exists
+        let (key, raw_blocks): (T::Key, RawValue<BlockNumberList>) = result.unwrap();
 
-        // If shard consists only of block numbers less than the target one, delete shard
-        // completely.
-        if key.as_ref().highest_block_number <= to_block {
-            cursor.delete_current()?;
-            deleted += 1;
-            if key.as_ref().highest_block_number == to_block {
-                // Shard contains only block numbers up to the target one, so we can skip to
-                // the last shard for this key. It is guaranteed that further shards for this
-                // sharded key will not contain the target block number, as it's in this shard.
-                cursor.seek_exact(last_key(&key))?;
+        // Get the highest block number that needs to be deleted for this sharded key
+        let to_block = sharded_key.as_ref().highest_block_number;
+
+        match prune_shard(&mut cursor, key, raw_blocks, to_block, &key_matches)? {
+            PruneShardOutcome::Deleted => deleted += 1,
+            PruneShardOutcome::Updated => updated += 1,
+            PruneShardOutcome::Unchanged => unchanged += 1,
+        }
+
+        while let Some((key, value)) = cursor
+            .next()?
+            .map(|(k, v)| Result::<_, DatabaseError>::Ok((k.key()?, v)))
+            .transpose()?
+        {
+            if key_matches(&key, &sharded_key) {
+                match prune_shard(&mut cursor, key, value, to_block, &key_matches)? {
+                    PruneShardOutcome::Deleted => deleted += 1,
+                    PruneShardOutcome::Updated => updated += 1,
+                    PruneShardOutcome::Unchanged => unchanged += 1,
+                }
+            } else {
+                break
             }
         }
```

Comment:
a lot of this code is redundant, it takes a while when reviewing your code to realise that different code blocks, actually do the same thing. addressing this as a suggestion was too big  https://github.com/paradigmxyz/reth/pull/9321